### PR TITLE
Jetty 12 mimetype cleanup

### PR DESCRIPTION
**Pull request by gregw**
Cleanup a bunch of TODOs related to MimeType lookup.
Primarily `MimeTypes` are now available via the `Context`.

### CODE REVIEW

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Response.java [hunk 1]
```diff
@@ -793,12 +793,12 @@ private String getCharacterEncoding(boolean setContentType)
             return _mimeType.getCharsetString();
 
         // Try charset assumed from content type (assumed charsets are not added to content type header).
-        encoding = MimeTypes.getCharsetAssumedFromContentType(_contentType);
+        encoding = _channel.getRequest().getCoreRequest().getContext().getMimeTypes().getCharsetAssumedFromContentType(_contentType);
```
Comment (lachlan-roberts) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Response.java]: Can we assign `_channel.getRequest().getCoreRequest().getContext().getMimeTypes()` to local variable.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java [hunk 2]
```diff
@@ -87,9 +88,15 @@ public static String getServerInfo()
         return "jetty/" + Server.getVersion();
     }
 
-    // TODO should persistent attributes be an Attributes.Layer over server attributes?
-    private final Attributes _persistentAttributes = new Mapped();
+    /*
+     * The context and specifically attributes and mimeTypes are not implemented as a layer over
+     * The server context, attributes and mimeTypes, because ultimately we may support cross context
```
Comment (lachlan-roberts) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java]: Maybe review this comment, could use some further clarification, and `The` should not be capitalised.

**Review by lachlan-roberts — CHANGES_REQUESTED** [review]

**Review by lachlan-roberts — APPROVED** [review]
LGTM

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java [hunk 3]
```diff
@@ -375,79 +247,188 @@ public static String getDefaultMimeByExtension(String filename)
                     break;
 
                 String ext = StringUtil.asciiToLowerCase(filename.substring(i + 1));
```
Comment (joakime) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java]: Use `FileID.getExtension(filename);`

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java [hunk 4]
```diff
@@ -222,145 +206,33 @@ public Type getBaseType()
...
     /**
      * Get the MIME type by filename extension.
-     * Lookup only the static default mime map.
      *
      * @param filename A file name
      * @return MIME type matching the longest dot extension of the
      * file name.
      */
-    public static String getDefaultMimeByExtension(String filename)
+    public String getMimeByExtension(String filename)
```
Comment (joakime) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java:236]: We should have a `getMimeByPath(Path path)` and which can use the fallback JVM mechanism.

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#probeContentType(java.nio.file.Path)

**Review by joakime — CHANGES_REQUESTED** [review]
Minor changes

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java (hunk 4 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java:236]: I think it is probably a good idea to use the prob somewhere in our resource handling, but I'm not sure that doing file ops in `MimeTypes` is the right way to go.   Maybe, but let's do that when we actually start using it. 

@lachlan-roberts can you think about this in your ResourceService work.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java [hunk 5]
```diff
@@ -202,7 +187,7 @@ public Type getBaseType()
         }
     }
 
-    public static final Index<Type> CACHE = new Index.Builder<Type>()
+    public static Index<Type> CACHE = new Index.Builder<Type>()
```
Comment (janbartel) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java]: why is this no longer final?

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java [hunk 6]
```diff
@@ -222,232 +207,215 @@ public Type getBaseType()
...
     }
 
-    /**
-     * Set a mime mapping
-     *
-     * @param extension the extension
-     * @param type the mime type
-     */
-    public void addMimeMapping(String extension, String type)
+    public static MimeTypes DEFAULTS = new MimeTypes(null)
```
Comment (janbartel) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java]: final?

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MimeTypesTest.java [hunk 7]
```diff
@@ -30,6 +30,8 @@ public static Stream<Arguments> mimeTypesByExtensionCases()
     {
         return Stream.of(
             Arguments.of("test.gz", "application/gzip"),
+            Arguments.of("test.tar.gz", "application/gzip"),
+            Arguments.of("test.tgz", "application/x-gtar"),
```
Comment (janbartel) [inline jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MimeTypesTest.java:34]: This is deliberate, yes? You removed type `application/x-gtar` from mime.properties ...

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java [hunk 8]
```diff
@@ -65,9 +65,23 @@ public static String getFileName(URI uri)
     }
 
     /**
+     * <p>
```
Comment (janbartel) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: Strictly doesn't have anything to do with `MimeTypes`, but seems like a good cleanup.

**Review by janbartel — APPROVED** [review]
Apart from a couple of finals etc. seems good.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java (hunk 8 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: yeah, this is actually a pre-merge of #8924 (pending it going Green and being merged).  So will not be a part of this PR once merged.

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MimeTypesTest.java (hunk 7 above)
Comment (gregw) [reply jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MimeTypesTest.java:34]: Correct.  Aligning with common extension handling practice of just looking at the final dot segment.  This is more efficient, same as most other projects and there is only one extension we had with a dot in it.

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Response.java [hunk 9]
```diff
@@ -792,13 +792,16 @@ private String getCharacterEncoding(boolean setContentType)
         if (_mimeType != null && _mimeType.isCharsetAssumed())
             return _mimeType.getCharsetString();
 
+        // Get MimeTypes from context
+        MimeTypes mimeTypes = _channel.getRequest().getCoreRequest().getContext().getMimeTypes();
```
Comment (joakime) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Response.java:796]: That's a long way to go, isn't the Context available already from this line?

**Review by joakime — APPROVED** [review]

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Response.java (hunk 9 above)
Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Response.java:796]: No.  Need to get the context and it is only available from the request.

**Pull request merged**